src/azul/openapi/validation.py

```diff
@@ -0,0 +1,183 @@
+                            relation_type = relation_schema['properties'][relation]['type']
+                        except ValueError:
+                            valid_relations = [list(x['properties'].keys())[0] for x in schema_property['oneOf']]
+                            raise BadRequestError(f"The relation in the `{request_parameter['name']}` parameter entry"
+                                                  f" for `{facet}` must be one of {valid_relations}")
+                        if not isinstance(value,
+                                          schema.type_lookup(relation_type)):
+                            raise BadRequestError(
+                                msg=f"The value of the `{relation}` relation in the `{request_parameter['name']}`"
+                                    f" parameter entry for `{facet}` is not a {relation_type}.")
+
+                except KeyError:
+                    raise BadRequestError(f"Unknown facet `{facet}`")
+
+        if getattr(self, "content") is not None:
+            _filter_validate(request_parameter)
+        else:
+            _schema_validate(request_parameter)
+
+
+def build_spec_validation(api_spec, resource_path, http_method) -> Mapping[str, List[SpecValidation]]:
+    spec_validators: Dict[str, List[SpecValidation]] = dict()
+    # _validate_requested_spec(api_spec, resource_path, http_method)
+    for parameter in _extract_parameters(api_spec, resource_path, http_method):
+        temp = SpecValidation(**parameter)
+        if 'schema' in parameter:
+            parameter_schema = parameter['schema']
+        elif 'content' in parameter:
+            for k, v in parameter['content'].items():
+                parameter_schema = v['schema']
+        else:
+            raise KeyError
+            # not sure if this is the correct error, is there validation done at this point?
+        temp.validator = schema.type_lookup(parameter_schema['type'])
+        spec_validators[temp.name] = temp
+    return spec_validators
+
+
+def validate_params(api_specs, current_request) -> None:
+    def fmt_error(err_description, parameters):
+        joined = ', '.join(f'`{p}`' for p in parameters)
+        return f'{err_description} {pluralize("parameter", len(parameters))} {joined}'
+
+    # OpenAPI requires HTTP method names be lower case
+    http_method = current_request.method.lower()
+    resource_path = current_request.context['resourcePath']
+
+    query_params = current_request.query_params if current_request.query_params else {}
+    uri_params = current_request.uri_params if current_request.uri_params else {}
+    all_provided_params = {**query_params, **uri_params}
+
+    provided_params = set(all_provided_params.keys())
+    mandatory_params = set()
+    validation_params = set()
+
+    spec_parameters = build_spec_validation(api_specs, resource_path, http_method)
+    for name, spec_parameter in spec_parameters.items():
+        if spec_parameter.required:
```

Comment:
PyCharm complains this is an `Unresolved attribute reference 'required' for class 'SpecValidation'`.
I see `SpecValidation` objects are created with dynamic arguments (`temp = SpecValidation(**parameter)`) however I feel since `required` is expected to always be defined it should be declared in the class.